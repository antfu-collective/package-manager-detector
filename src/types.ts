export type Agent = 'npm' | 'yarn' | 'yarn@berry' | 'pnpm' | 'pnpm@6' | 'bun'

export type AgentCommandValue = (string | number)[] | ((args?: string[]) => string[]) | null

export interface AgentCommands {
  'agent': AgentCommandValue
  'run': AgentCommandValue
  'install': AgentCommandValue
  'frozen': AgentCommandValue
  'global': AgentCommandValue
  'add': AgentCommandValue
  'upgrade': AgentCommandValue
  'upgrade-interactive': AgentCommandValue
  'execute': AgentCommandValue
  'uninstall': AgentCommandValue
  'global_uninstall': AgentCommandValue
}

export type Command = keyof AgentCommands

export interface CommandType {
  command: string
  arguments: string[]
  toString: () => string
}

export type CommandReturnType = (args?: string[]) => CommandType

export interface ResolvedCommand {
  command: string
  args: string[]
}
