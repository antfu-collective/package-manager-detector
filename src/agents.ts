export const AGENTS = ['npm', 'yarn', 'yarn@berry', 'pnpm', 'pnpm@6', 'bun'] as const
export type Agent = typeof AGENTS[number]

// the order here matters, more specific one comes first
export const LOCKS: Record<string, Agent> = {
  'bun.lockb': 'bun',
  'pnpm-lock.yaml': 'pnpm',
  'yarn.lock': 'yarn',
  'package-lock.json': 'npm',
  'npm-shrinkwrap.json': 'npm',
}

function npmRun(agent: string) {
  return (args: string[]) => {
    if (args.length > 1)
      return `${agent} run ${args[0]} -- ${args.slice(1).join(' ')}`
    else return `${agent} run ${args[0]}`
  }
}

const yarn = {
  'agent': 'yarn {0}',
  'run': 'yarn run {0}',
  'install': 'yarn install {0}',
  'frozen': 'yarn install --frozen-lockfile',
  'global': 'yarn global add {0}',
  'add': 'yarn add {0}',
  'upgrade': 'yarn upgrade {0}',
  'upgrade-interactive': 'yarn upgrade-interactive {0}',
  'execute': 'npx {0}',
  'uninstall': 'yarn remove {0}',
  'global_uninstall': 'yarn global remove {0}',
}
const pnpm = {
  'agent': 'pnpm {0}',
  'run': 'pnpm run {0}',
  'install': 'pnpm i {0}',
  'frozen': 'pnpm i --frozen-lockfile',
  'global': 'pnpm add -g {0}',
  'add': 'pnpm add {0}',
  'upgrade': 'pnpm update {0}',
  'upgrade-interactive': 'pnpm update -i {0}',
  'execute': 'pnpm dlx {0}',
  'uninstall': 'pnpm remove {0}',
  'global_uninstall': 'pnpm remove --global {0}',
}
const bun = {
  'agent': 'bun {0}',
  'run': 'bun run {0}',
  'install': 'bun install {0}',
  'frozen': 'bun install --frozen-lockfile',
  'global': 'bun add -g {0}',
  'add': 'bun add {0}',
  'upgrade': 'bun update {0}',
  'upgrade-interactive': 'bun update {0}',
  'execute': 'bun x {0}',
  'uninstall': 'bun remove {0}',
  'global_uninstall': 'bun remove -g {0}',
}

export const COMMANDS = {
  'npm': {
    'agent': 'npm {0}',
    'run': npmRun('npm'),
    'install': 'npm i {0}',
    'frozen': 'npm ci',
    'global': 'npm i -g {0}',
    'add': 'npm i {0}',
    'upgrade': 'npm update {0}',
    'upgrade-interactive': null,
    'execute': 'npx {0}',
    'uninstall': 'npm uninstall {0}',
    'global_uninstall': 'npm uninstall -g {0}',
  },
  'yarn': yarn,
  'yarn@berry': {
    ...yarn,
    'frozen': 'yarn install --immutable',
    'upgrade': 'yarn up {0}',
    'upgrade-interactive': 'yarn up -i {0}',
    'execute': 'yarn dlx {0}',
    // Yarn 2+ removed 'global', see https://github.com/yarnpkg/berry/issues/821
    'global': 'npm i -g {0}',
    'global_uninstall': 'npm uninstall -g {0}',
  },
  'pnpm': pnpm,
  // pnpm v6.x or below
  'pnpm@6': {
    ...pnpm,
    run: npmRun('pnpm'),
  },
  'bun': bun,
}

export type Command = keyof typeof COMMANDS.npm

export interface CommandType {
  command: string
  arguments: string[]
  toString: () => string
}

export type CommandReturnType = (args?: string[]) => CommandType
export interface AgentCommand {
  'agent': string
  'run': CommandReturnType
  'install': CommandReturnType
  'frozen': CommandReturnType
  'global': CommandReturnType
  'add': CommandReturnType
  'upgrade': CommandReturnType
  'upgrade-interactive'?: CommandReturnType
  'execute': CommandReturnType
  'uninstall': CommandReturnType
  'global_uninstall': CommandReturnType
}

function npmRunCommand(agent: string) {
  return (args: string[] = []) => {
    return {
      command: agent,
      arguments: args.length > 1
        ? ['run', args[0], '--', ...args.slice(1)]
        : ['run', args[0]],
      toString() {
        return [agent, ...this.arguments].join(' ')
      },
    } satisfies CommandType
  }
}

function quoteArgument(arg: string) {
  return (!arg.startsWith('--') && arg.includes(' '))
    ? JSON.stringify(arg)
    : arg
}

function buildCommand(agentCommand: string) {
  const [command, ...commandArgs] = agentCommand.split(' ')
  const last = commandArgs.at(-1) === '{0}'
  if (last) {
    commandArgs.pop()
  }

  return (args: string[] = []) => {
    return {
      command,
      arguments: last ? [...commandArgs, ...args.map(quoteArgument)] : [...commandArgs],
      toString() {
        return [command, ...this.arguments].join(' ')
      },
    } satisfies CommandType
  }
}

function buildCommands(agent: Agent, commands: typeof COMMANDS.npm) {
  const interactive = commands['upgrade-interactive'] as string | null
  return {
    'agent': commands.agent,
    'run': typeof commands.run === 'function'
      ? npmRunCommand(agent.includes('@') ? agent.split('@')[0] : agent)
      : buildCommand(commands.run as string),
    'install': buildCommand(commands.install),
    'frozen': buildCommand(commands.frozen),
    'global': buildCommand(commands.global),
    'add': buildCommand(commands.add),
    'upgrade': buildCommand(commands.upgrade),
    'upgrade-interactive': interactive ? buildCommand(interactive) : undefined,
    'execute': buildCommand(commands.execute),
    'uninstall': buildCommand(commands.uninstall),
    'global_uninstall': buildCommand(commands.global_uninstall),
  } satisfies AgentCommand
}

export const AGENT_COMMANDS = Object.entries(COMMANDS)
  .map(([pm, commands]) => [pm, buildCommands(pm as Agent, commands as any)] as const)
  .reduce((acc, [pm, commands]) => {
    acc[pm as Agent] = commands
    return acc
  }, {} as Record<Agent, ReturnType<typeof buildCommands>>)

export const INSTALL_PAGE: Record<Agent, string> = {
  'bun': 'https://bun.sh',
  'pnpm': 'https://pnpm.io/installation',
  'pnpm@6': 'https://pnpm.io/6.x/installation',
  'yarn': 'https://classic.yarnpkg.com/en/docs/install',
  'yarn@berry': 'https://yarnpkg.com/getting-started/install',
  'npm': 'https://docs.npmjs.com/cli/v8/configuring-npm/install',
}
