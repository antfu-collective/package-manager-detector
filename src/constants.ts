import type { Agent, AgentName } from './types'

export const AGENTS: Agent[] = [
  'npm',
  'yarn',
  'yarn@berry',
  'pnpm',
  'pnpm@6',
  'bun',
]

// the order here matters, more specific one comes first
export const LOCKS: Record<string, AgentName> = {
  'bun.lockb': 'bun',
  'pnpm-lock.yaml': 'pnpm',
  'yarn.lock': 'yarn',
  'package-lock.json': 'npm',
  'npm-shrinkwrap.json': 'npm',
}

export const INSTALL_PAGE: Record<Agent, string> = {
  'bun': 'https://bun.sh',
  'pnpm': 'https://pnpm.io/installation',
  'pnpm@6': 'https://pnpm.io/6.x/installation',
  'yarn': 'https://classic.yarnpkg.com/en/docs/install',
  'yarn@berry': 'https://yarnpkg.com/getting-started/install',
  'npm': 'https://docs.npmjs.com/cli/v8/configuring-npm/install',
}
