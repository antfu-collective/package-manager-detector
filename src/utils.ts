export const AGENTS = ['npm', 'yarn', 'yarn@berry', 'pnpm', 'pnpm@6', 'bun'] as const
export type Agent = typeof AGENTS[number]

// the order here matters, more specific one comes first
export const LOCKS: Record<string, Agent> = {
  'bun.lockb': 'bun',
  'pnpm-lock.yaml': 'pnpm',
  'yarn.lock': 'yarn',
  'package-lock.json': 'npm',
  'npm-shrinkwrap.json': 'npm',
}
