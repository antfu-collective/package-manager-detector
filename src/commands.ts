import type { Agent } from './agents'
import { COMMANDS } from './agents'

export type CommandType = [command: string, args: string[]]
export interface AgentCommand {
  'agent': string
  'run': CommandType | ((args: string[]) => CommandType)
  'install': CommandType
  'frozen': CommandType
  'global': CommandType
  'add': CommandType
  'upgrade': CommandType
  'upgrade-interactive': CommandType | null
  'execute': CommandType
  'uninstall': CommandType
  'global_uninstall': CommandType
}

function npmRun(agent: string) {
  return (args: string[]): CommandType => {
    return args.length > 1
      ? [agent, ['run', args[0], '--', ...args.slice(1)]]
      : [agent, ['run', args[0]]]
  }
}

function buildCommand(command: string) {
  const [cmd, ...args] = command.split(' ')
  return [cmd, args] as CommandType
}

function buildCommands(agent: Agent, commands: typeof COMMANDS.npm) {
  return {
    'agent': commands.agent,
    'run': typeof commands.run === 'function'
      ? npmRun(agent.includes('@') ? agent.split('@')[0] : agent)
      : buildCommand(commands.run),
    'install': buildCommand(commands.install),
    'frozen': buildCommand(commands.frozen),
    'global': buildCommand(commands.global),
    'add': buildCommand(commands.add),
    'upgrade': buildCommand(commands.upgrade),
    'upgrade-interactive': commands['upgrade-interactive']
      ? buildCommand(commands['upgrade-interactive'])
      : null,
    'execute': buildCommand(commands.execute),
    'uninstall': buildCommand(commands.uninstall),
    'global_uninstall': buildCommand(commands.global_uninstall),
  } satisfies AgentCommand
}

export const AGENT_COMMANDS = Object.entries(COMMANDS)
  .map(([pm, commands]) => [pm, buildCommands(pm as Agent, commands as any)] as const)
  .reduce((acc, [pm, commands]) => {
    acc[pm as Agent] = buildCommands(pm as Agent, commands as any)
    return acc
  }, {} as Record<Agent, ReturnType<typeof buildCommands>>)
