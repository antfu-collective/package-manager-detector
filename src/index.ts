import fs from 'node:fs'
import fsPromises from 'node:fs/promises'
import path from 'node:path'
import process from 'node:process'
import type { Agent } from './agents'
import { AGENTS, LOCKS } from './agents'

export interface DetectOptions {
  cwd?: string
  /**
   * Callback when unknown package manager from package.json.
   *
   * @param packageManager - The `packageManager` value from package.json file.
   */
  onUnknown?: (packageManager: string) => DetectResult | null | undefined
}

export type { Agent }
export { AGENTS, LOCKS }

export interface DetectResult {
  agent: Agent
  version?: string
}

export async function detect({ cwd, onUnknown }: DetectOptions = {}): Promise<DetectResult | null> {
  let agent: Agent | undefined

  for (const directory of lookup(cwd)) {
    // Look up for lock files
    for (const lock of Object.keys(LOCKS)) {
      if (await fileExists(path.join(directory, lock))) {
        agent = LOCKS[lock]
        const result = await parsePackageJson(path.join(directory, 'package.json'), onUnknown)
        if (result)
          return result
        else
          return { agent }
      }
    }
    // Look up for package.json
    const result = await parsePackageJson(path.join(directory, 'package.json'), onUnknown)
    if (result)
      return result
  }

  return null
}

function * lookup(cwd: string = process.cwd()): Generator<string> {
  let directory = path.resolve(cwd)
  const { root } = path.parse(directory)

  while (directory && directory !== root) {
    yield directory

    directory = path.dirname(directory)
  }
}

async function parsePackageJson(
  filepath: string,
  onUnknown: DetectOptions['onUnknown'],
): Promise<DetectResult | null> {
  // read `packageManager` field in package.json
  if (!filepath || !await fileExists(filepath))
    return null

  try {
    const pkg = JSON.parse(fs.readFileSync(filepath, 'utf8'))
    let agent: Agent | undefined
    if (typeof pkg.packageManager === 'string') {
      const [name, ver] = pkg.packageManager.replace(/^\^/, '').split('@')
      let version = ver
      if (name === 'yarn' && Number.parseInt(ver) > 1) {
        agent = 'yarn@berry'
        // the version in packageManager isn't the actual yarn package version
        version = 'berry'
        return { agent, version }
      }
      else if (name === 'pnpm' && Number.parseInt(ver) < 7) {
        agent = 'pnpm@6'
        return { agent, version }
      }
      else if (AGENTS.includes(name)) {
        agent = name as Agent
        return { agent, version }
      }
      else {
        return onUnknown?.(pkg.packageManager) ?? null
      }
    }
  }
  catch {}
  return null
}

async function fileExists(filePath: string) {
  try {
    const stats = await fsPromises.stat(filePath)
    if (stats.isFile()) {
      return true
    }
  }
  catch {}
  return false
}
