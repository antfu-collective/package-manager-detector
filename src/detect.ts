import fs from 'node:fs'
import fsPromises from 'node:fs/promises'
import path from 'node:path'
import process from 'node:process'
import type { Agent, DetectOptions, DetectResult } from './types'
import { AGENTS, LOCKS } from './constants'

export async function detect({ cwd, onUnknown }: DetectOptions = {}): Promise<DetectResult | null> {
  let agent: Agent | undefined

  for (const directory of lookup(cwd)) {
    // Look up for lock files
    for (const lock of Object.keys(LOCKS)) {
      if (await fileExists(path.join(directory, lock))) {
        agent = LOCKS[lock]
        const result = await parsePackageJson(path.join(directory, 'package.json'), onUnknown)
        if (result)
          return result
        else
          return { agent }
      }
    }
    // Look up for package.json
    const result = await parsePackageJson(path.join(directory, 'package.json'), onUnknown)
    if (result)
      return result
  }

  return null
}

function * lookup(cwd: string = process.cwd()): Generator<string> {
  let directory = path.resolve(cwd)
  const { root } = path.parse(directory)

  while (directory && directory !== root) {
    yield directory

    directory = path.dirname(directory)
  }
}

async function parsePackageJson(
  filepath: string,
  onUnknown: DetectOptions['onUnknown'],
): Promise<DetectResult | null> {
  // read `packageManager` field in package.json
  if (!filepath || !await fileExists(filepath))
    return null

  try {
    const pkg = JSON.parse(fs.readFileSync(filepath, 'utf8'))
    let agent: Agent | undefined
    if (typeof pkg.packageManager === 'string') {
      const [name, ver] = pkg.packageManager.replace(/^\^/, '').split('@')
      let version = ver
      if (name === 'yarn' && Number.parseInt(ver) > 1) {
        agent = 'yarn@berry'
        // the version in packageManager isn't the actual yarn package version
        version = 'berry'
        return { agent, version }
      }
      else if (name === 'pnpm' && Number.parseInt(ver) < 7) {
        agent = 'pnpm@6'
        return { agent, version }
      }
      else if (AGENTS.includes(name)) {
        agent = name as Agent
        return { agent, version }
      }
      else {
        return onUnknown?.(pkg.packageManager) ?? null
      }
    }
  }
  catch {}
  return null
}

async function fileExists(filePath: string) {
  try {
    const stats = await fsPromises.stat(filePath)
    if (stats.isFile()) {
      return true
    }
  }
  catch {}
  return false
}
