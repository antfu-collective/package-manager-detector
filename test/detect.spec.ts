import type { MockInstance } from 'vitest'
import type { DetectOptions } from '../src'
import { tmpdir } from 'node:os'
import path from 'node:path'
import fs from 'fs-extra'
import { afterAll, beforeAll, describe, expect, it, vi } from 'vitest'
import { detect } from '../src'

let basicLog: MockInstance, errorLog: MockInstance, warnLog: MockInstance, infoLog: MockInstance

function detectTest(fixture: string, agent: string, options?: DetectOptions) {
  return async () => {
    const cwd = await fs.mkdtemp(path.join(tmpdir(), 'ni-'))
    const dir = path.join(__dirname, 'fixtures', fixture, agent)
    await fs.copy(dir, cwd)

    expect(await detect({ cwd, ...options })).toMatchSnapshot()
  }
}

beforeAll(() => {
  basicLog = vi.spyOn(console, 'log')
  warnLog = vi.spyOn(console, 'warn')
  errorLog = vi.spyOn(console, 'error')
  infoLog = vi.spyOn(console, 'info')
})

afterAll(() => {
  vi.resetAllMocks()
})

const fixtures = ['lockfile', 'packager', 'dev-engines', 'install-metadata']

fixtures.forEach(fixture => describe(fixture, () => {
  const fixtureDirs = getFixtureDirs(fixture)

  fixtureDirs.forEach((dir) => {
    let options: DetectOptions | undefined
    if (fixture === 'install-metadata') {
      options = { strategies: ['install-metadata', 'lockfile', 'packageManager-field', 'devEngines-field'] }
    }
    it(dir, detectTest(fixture, dir, options))
  })

  it('no logs', () => {
    expect(basicLog).not.toHaveBeenCalled()
    expect(warnLog).not.toHaveBeenCalled()
    expect(errorLog).not.toHaveBeenCalled()
    expect(infoLog).not.toHaveBeenCalled()
  })
}))

function getFixtureDirs(fixture: string) {
  const fixtureDir = path.join(__dirname, 'fixtures', fixture)
  const items = fs.readdirSync(fixtureDir)
  return items.filter(item => fs.statSync(path.join(fixtureDir, item)).isDirectory())
}

it('stops at specified directory', async () => {
  const cwd = await fs.mkdtemp(path.join(tmpdir(), 'ni-'))

  const noFilesDir = path.join(cwd, 'no-files')
  const nestedNoFilesDir = path.join(noFilesDir, 'nested')
  const parentDir = cwd

  await fs.mkdirp(noFilesDir)
  await fs.mkdirp(nestedNoFilesDir)

  await fs.copy(
    path.join(__dirname, 'fixtures', 'lockfile', 'npm'),
    parentDir,
  )

  const resultWithStop = await detect({
    cwd: nestedNoFilesDir,
    stop: noFilesDir,
  })

  expect(resultWithStop).toBe(null)

  const resultWithoutStop = await detect({
    cwd: nestedNoFilesDir,
  })

  expect(resultWithoutStop).toMatchObject({
    name: 'npm',
    agent: 'npm',
  })
})
