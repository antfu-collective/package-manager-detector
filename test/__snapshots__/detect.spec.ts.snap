// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`lockfile > bun 1`] = `
{
  "agent": "bun",
}
`;

exports[`lockfile > npm 1`] = `
{
  "agent": "npm",
}
`;

exports[`lockfile > pnpm 1`] = `
{
  "agent": "pnpm",
}
`;

exports[`lockfile > pnpm@6 1`] = `
{
  "agent": "pnpm",
}
`;

exports[`lockfile > unknown 1`] = `null`;

exports[`lockfile > yarn 1`] = `
{
  "agent": "yarn",
}
`;

exports[`lockfile > yarn@berry 1`] = `
{
  "agent": "yarn",
}
`;

exports[`packager > bun 1`] = `
{
  "agent": "bun",
  "version": "0",
}
`;

exports[`packager > npm 1`] = `
{
  "agent": "npm",
  "version": "7",
}
`;

exports[`packager > pnpm 1`] = `
{
  "agent": "pnpm",
  "version": "8",
}
`;

exports[`packager > pnpm@6 1`] = `
{
  "agent": "pnpm@6",
  "version": "6",
}
`;

exports[`packager > unknown 1`] = `null`;

exports[`packager > yarn 1`] = `
{
  "agent": "yarn",
  "version": "1",
}
`;

exports[`packager > yarn@berry 1`] = `
{
  "agent": "yarn@berry",
  "version": "berry",
}
`;
